### PR TITLE
Fix: Add max_docs_per_bin and best-fit for optimal pre-training

### PR DESCRIPTION
**Pull request by dtamayo-nlp**
## Motivation

xdoc attention masking becomes inefficient when samples contain many very small documents. Without xdoc attention masking, samples are uniformly sized, each parallel group processes sequences equivalent lengths, leading to well-balanced workloads and straightforward synchronization.

However, with xdoc masking, pathological cases can arise. For example, a sequence of 8192 tokens may contain hundreds of tiny documents (e.g., ~410 documents of ~20 tokens each). This forces the model to repeatedly process very small attention contexts, significantly increasing overhead.

Even if such cases are rare, the training step must remain synchronized across all parallel workers. As a result, these outliers determine the step time and degrade overall throughput.

## Problem Visualization

Throughput becomes highly unstable over time when these pathological samples are encountered, as shown below:
image

## Solution

Introduce a cap on the number of documents per sample. When a sample contains many very small documents, we truncate the number of documents to a predefined limit and pad the remaining sequence length.

In practice, these edge cases are rare (observed in only 4 datasets for the 70B setting), so the additional padding overhead is minimal while significantly improving stability and throughput.

## Changes

Add a [best-fit decreasing (BFD)](https://arxiv.org/pdf/2404.10830) packing strategy with support for limiting the number of documents per sample.

This change is fully backward compatible with the main branch if these arguments are not given. The new behavior is enabled via:

```
--pretraining-packing-strategy bfd
--max-docs-per-bin 16
```

which activates document-capped packing.

### CODE REVIEW

**Comment by dtamayo-nlp** [comment]
## Trade-off between `max-docs-per-bin` and throughput

Reducing the maximum number of documents per bin leads to more stable throughput, but at the cost of increased padding.

We evaluated caps of 16 and 32 documents per bin:

**Packing efficiency**

* 16 docs/bin: ~90% average efficiency (~10% padding). Most datasets exceeded 95%, though 1146/3624 fell below 90%.
* 32 docs/bin: About half the datasets reached 99% efficiency, and 3576/3624 exceeded 90%.

**Throughput**

* With higher limits (e.g., 32 docs/bin), throughput exhibited slightly greater variance moving around 620 and 680 tokens/gpu/s. This effect may diminish after a few iterations.

image

**Comment by dtamayo-nlp** [comment]
Closed given that the throughput became unstable after some iterations, the cause is still not clear.
